File: source/winx86/CommLayer/NetUser.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Collections;
using System.Net;
using System.Net.Sockets;
using NetLayer;

namespace CommLayer
{
    /// <summary>
    /// Clase para enumeración de posibles calidades de señal de los usuarios remotos
    /// </summary>
    public static class NetUserQuality
    {
        /// <summary>
        /// Constante para el estado normal
        /// </summary>
        public const Int32 NORMAL = 1;

        /// <summary>
        /// Constante para el estado bajo
        /// </summary>
        public const Int32 LOW = 2;

        /// <summary>
        /// Constante para el estado critico
        /// </summary>
        public const Int32 CRITICAL = 3;
    }

    /// <summary>
    /// Clase para enumeración de posibles estados de tráfico de los usuarios remotos
    /// </summary>
    public static class CommunicationQuality
    {
        /// <summary>
        /// Constante para el estado normal
        /// </summary>
        public const int NORMAL = 1;

        /// <summary>
        /// Constante para el estado con sobrecargar de trafico
        /// </summary>
        public const int OVERLOADED = 2;

        /// <summary>
        /// Constante para el estado critico
        /// </summary>
        public const int CRITICAL = 3;
    }


    /// <summary>
    /// Datos de usuario dentro de la MANET
    /// </summary>
    public class NetUser
    {
        /// <summary>
        /// Id del usuario, esta variable se serializa
        /// </summary>
        private Guid _id;

        /// <summary>
        /// Nombre del usuario, esta variable se serializa
        /// </summary>
        private String _name;

        /// <summary>
        /// Dirección IP, esta variable se serializa
        /// </summary>
        private IPAddress _ip;

        /// <summary>
        /// Array con los id's de la vecindad TCP, esta variable se serializa
        /// </summary>
        private Guid[] _neighborhoodIds;

        /// <summary>
        /// La calidad de la señal, NO se serializa
        /// </summary>
        private Int32 _signalQuality;

        /// <summary>
        /// Valor de señal restante, NO se serializa
        /// </summary>
        private Int32 _timeout;

        /// <summary>
        /// Saltos UDP que se encuentra del usuario remoto, NO se serializa
        /// </summary>
        private Int32 _jumpsAway;

        /// <summary>
        /// Tiempo de espera restante para intentar una conexion TCP con este usuario remoto, NO se serializa
        /// </summary>
        private Int32 _waitTimeOut;

        /// <summary>
        /// El estado de la señal
        /// </summary>
        private Int32 _state;

        /// <summary>
        /// El grupo al que pertenece
        /// </summary>
        private byte[] _upLayerData;

        /// <summary>
        /// Default Constructor
        /// </summary>
        public NetUser()
        {
            this.Name = System.Environment.MachineName;
            this.SignalQuality = NetUserQuality.NORMAL;
            this.NeighborhoodIds = new Guid[0];
            this.JumpsAway = 0;
            this.WaitTimeOut = 0;
            this.State = CommunicationQuality.NORMAL;
            this.UpLayerData = new byte[0];
        }

        /// <summary>
        /// Constructor parametrizado, para uso interno, no se debe usar directamente
        /// </summary>
        /// <param name="id">El id del usuario</param>
        /// <param name="name">El nombre del usuario</param>
        /// <param name="ip">La dirección IP del usuario</param>
        /// <param name="neighborhood">Un array con los ids de los nodos vecinos al usuario</param>
        /// <param name="netData">Los datos de configuración</param>
        internal NetUser(Guid id, String name, IPAddress ip, Guid[] neighborhood, NetData netData)
        {
            this.Id = id;
            this.Name = name;
            this.Ip = ip;
            this.NeighborhoodIds = neighborhood;
            this.SignalQuality = NetUserQuality.NORMAL;
            setTimeOut(netData.QualityMaxNetUser, netData);
            this.JumpsAway = 0;
            this.WaitTimeOut = 0;
            this.State = CommunicationQuality.NORMAL;
            this.UpLayerData = new byte[0];
        }

        /// <summary>
        /// Nombre a mostrar
        /// </summary>
        public String Name
        {
            get { return _name; }
            set { _name = value; }
        }

        /// <summary>
        /// Dirección Ip
        /// </summary>
        public IPAddress Ip
        {
            get { return _ip; }
            set { _ip = value; }
        }

        /// <summary>
        /// Calidad de la señal
        /// </summary>
        public Int32 SignalQuality
        {
            get { return _signalQuality; }
            set { _signalQuality = value; }
        }

        /// <summary>
        /// El numero único de indentificación del usuario
        /// </summary>
        public Guid Id
        {
            get { return _id; }
            set { _id = value; }
        }

        /// <summary>
        /// Los nodos vecinos del usuario (array de los ids de los usuarios con conexión directa TCP)
        /// </summary>
        public Guid[] NeighborhoodIds
        {
            get { return _neighborhoodIds; }
            set { _neighborhoodIds = value; }
        }

        /// <summary>
        /// Tiempo restante para que la calidad de la señal llegue a 0
        /// </summary>
        public Int32 Timeout
        {
            get { return _timeout; }
        }

        /// <summary>
        /// Cantidad de saltos UDP que se encuentra el usuario remoto
        /// </summary>
        public Int32 JumpsAway
        {
            get { return _jumpsAway; }
            set { _jumpsAway = value; }
        }

        /// <summary>
        /// Tiempo de espera para intentar conectar por TCP
        /// </summary>
        internal Int32 WaitTimeOut
        {
            get { return _waitTimeOut; }
            set { _waitTimeOut = value; }
        }

        /// <summary>
        /// La calidad del estado de comunicación (señal)
        /// </summary>
        public Int32 State
        {
            get { return _state; }
            set { _state = value; }
        }

        /// <summary>
        /// El grupo al que pertenece
        /// </summary>
        public byte[] UpLayerData
        {
            get { return _upLayerData; }
            set { _upLayerData = value; }
        }

        /// <summary>
        /// Chequea que el usuario ha sido detectado, incrementando su valor de calidad
        /// Método para uso interno
        /// </summary>
        /// <param name="netData">Los parámetros de configuración de red</param>
        internal void qualityUp(NetData netData)
        {
            setTimeOut(Timeout + netData.QualityRiseNetUser, netData);
        }

        /// <summary>
        /// Decrementa la calidad de señal, método para uso interno
        /// </summary>
        internal void qualityDown(NetData netData)
        {
            setTimeOut(Timeout - 1, netData);
        }

        /// <summary>
        /// Setea el time Out del este usuario para medir calidad de señal
        /// </summary>
        /// <param name="newTimeOut">El nuevo timeout</param>
        /// <param name="netData">Los datos de configuración de red</param>
        internal void setTimeOut(Int32 newTimeOut, NetData netData) 
        {
            if (newTimeOut > netData.QualityMaxNetUser)
            {
                _timeout = netData.QualityMaxNetUser;
            }
            else
            {
                _timeout = newTimeOut;
            }

            if (_timeout > netData.QualityNormalNetUser)
            {
                this.SignalQuality = NetUserQuality.NORMAL;
            }
            else if (_timeout > netData.QualityLowNetUser)
            {
                this.SignalQuality = NetUserQuality.LOW;
            }
            else
            {
                this.SignalQuality = NetUserQuality.CRITICAL;
            }
        }

        /// <summary>
        /// Selecciona un nuevo ID, método para uso interno
        /// </summary>
        internal void pickNewId()
        {
            Id = Guid.NewGuid();
        }

        /// <summary>
        /// Fabrica una representación del usuario como String
        /// </summary>
        /// <returns>El nombre del usuario y su numero IP</returns>
        public override string ToString()
        {
            return Name;
        }
    }
}

File: source/java/hlmp/CommLayer/NetUser.java
package hlmp.CommLayer;

import hlmp.CommLayer.Constants.*;
import hlmp.NetLayer.NetData;

import java.net.InetAddress;
import java.net.UnknownHostException;
import java.util.UUID;

/**
 * Datos de usuario dentro de la MANET
 */
public class NetUser {

	/**
     * Id del usuario, esta variable se serializa
     */
    private UUID id;

    /**
     * Nombre del usuario, esta variable se serializa
     */
    private String name;

    /**
     * Dirección IP, esta variable se serializa
     */
    private InetAddress ip;

    /**
     * Array con los id's de la vecindad TCP, esta variable se serializa
     */
    private UUID[] neighborhoodIds;

    /**
     * La calidad de la señal, NO se serializa
     */
    private int signalQuality;

    /**
     * Valor de señal restante, NO se serializa
     */
    private int timeout;

    /**
     * Saltos UDP que se encuentra del usuario remoto, NO se serializa
     */
    private int jumpsAway;

    /**
     * Tiempo de espera restante para intentar una conexion TCP con este usuario remoto, NO se serializa
     */
    private int waitTimeOut;

    /**
     * El estado de la señal
     */
    private int state;

    /**
     * El grupo al que pertenece
     */
    private byte[] upLayerData;
    
    /**
     * Default Constructor
     */
    public NetUser()
    {
        try {
			this.name = InetAddress.getLocalHost().getHostName();
		} catch (UnknownHostException e) {
			e.printStackTrace();
			this.name = "undef";
		}
        this.signalQuality = NetUserQuality.NORMAL;
        this.neighborhoodIds = new UUID[0];
        this.jumpsAway = 0;
        this.waitTimeOut = 0;
        this.state = CommunicationQuality.NORMAL;
        this.upLayerData = new byte[0];
    }
    
    /**
     * Constructor parametrizado, para uso interno, no se debe usar directamente
     * @param id El id del usuario
     * @param name El nombre del usuario
     * @param ip La dirección IP del usuario
     * @param neighborhood Un array con los ids de los nodos vecinos al usuario
     * @param netData Los datos de configuración
     */
    public NetUser(UUID id, String name, InetAddress ip, UUID[] neighborhood, NetData netData)
    {
        this.id = id;
        this.name = name;
        this.ip = ip;
        this.neighborhoodIds = neighborhood;
        this.signalQuality = NetUserQuality.NORMAL;
        setTimeOut(netData.getQualityMaxNetUser(), netData);
        this.jumpsAway = 0;
        this.waitTimeOut = 0;
        this.state = CommunicationQuality.NORMAL;
        this.upLayerData = new byte[0];
    }
    
    /**
     * Chequea que el usuario ha sido detectado, incrementando su valor de calidad
     * Método para uso interno
     * @param netData Los parámetros de configuración de red
     */
    public void qualityUp(NetData netData)
    {
        setTimeOut(this.timeout + netData.getQualityRiseNetUser(), netData);
    }

    /**
     * Decrementa la calidad de señal, método para uso interno
     */
    public void qualityDown(NetData netData)
    {
        setTimeOut(this.timeout - 1, netData);
    }
    
    /**
     * Setea el time Out del este usuario para medir calidad de señal
     * @param newTimeOut El nuevo timeout
     * @param netData Los datos de configuración de red
     */
    private void setTimeOut(int newTimeOut, NetData netData) 
    {
        if (newTimeOut > netData.getQualityMaxNetUser())
        {
            this.timeout = netData.getQualityMaxNetUser();
        }
        else
        {
            this.timeout = newTimeOut;
        }

        if (this.timeout > netData.getQualityNormalNetUser())
        {
            this.signalQuality = NetUserQuality.NORMAL;
        }
        else if (this.timeout > netData.getQualityLowNetUser())
        {
            this.signalQuality = NetUserQuality.LOW;
        }
        else
        {
            this.signalQuality = NetUserQuality.CRITICAL;
        }
    }
	
    /**
     * Selecciona un nuevo ID, método para uso interno
     */
    public void pickNewId()
    {
        this.id = UUID.randomUUID();
    }

    /**
     * Fabrica una representación del usuario como String
     * @return El nombre del usuario y su numero IP
     */
    @Override
    public String toString()
    {
        return this.name + " ("+this.ip.getHostAddress()+")";
    }

	public UUID getId() {
		return id;
	}

	public void setId(UUID id) {
		this.id = id;
	}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public InetAddress getIp() {
		return ip;
	}

	public void setIp(InetAddress ip) {
		this.ip = ip;
	}

	public UUID[] getNeighborhoodIds() {
		return neighborhoodIds;
	}

	public void setNeighborhoodIds(UUID[] neighborhoodIds) {
		this.neighborhoodIds = neighborhoodIds;
	}
	
	public void setNeighborhoodId(int pos, UUID neighborhoodId){
		this.neighborhoodIds[pos] = neighborhoodId;
	}

	public int getSignalQuality() {
		return signalQuality;
	}

	public void setSignalQuality(int signalQuality) {
		this.signalQuality = signalQuality;
	}

	/**
	 * @return Valor de señal restante
	 */
	public int getTimeout() {
		return timeout;
	}

//	public void setTimeout(int timeout) {
//		this.timeout = timeout;
//	}

	public int getJumpsAway() {
		return jumpsAway;
	}

	public void setJumpsAway(int jumpsAway) {
		this.jumpsAway = jumpsAway;
	}

	/**
	 * @return Tiempo de espera restante para intentar una conexion TCP con este usuario remoto
	 */
	public int getWaitTimeOut() {
		return waitTimeOut;
	}

	public void setWaitTimeOut(int waitTimeOut) {
		this.waitTimeOut = waitTimeOut;
	}
	
	public void waitTimeOutDown(){
		this.waitTimeOut--;
	}

	public int getState() {
		return state;
	}

	public void setState(int state) {
		this.state = state;
	}

	public byte[] getUpLayerData() {
		return upLayerData;
	}

	public void setUpLayerData(byte[] upLayerData) {
		this.upLayerData = upLayerData;
	}

}